File: 1주차/simsimjae/ValidParentheses.js
var isValid = function(s) {
    const stack = [];

    const isPair = (stackTop, char) => {
        if(stackTop === '(' && char === ')') return true;
        else if(stackTop === '[' && char === ']') return true;
        else if(stackTop === '{' && char === '}') return true;
        else return false;
    }
    
    for(let i=0; i<s.length; i++) {
        if(!s.length) return true;
        
        const char = s[i];
        const stackTop = stack.length > 0 && stack[stack.length - 1];
        
        if(!stack.length) {
            stack.push(char);    
        }else if(isPair(stackTop, char)) {
            stack.pop();
        }else {
            stack.push(char);
        }
    }
    
    
    return stack.length === 0;
};

File: 1주차/kmdngyu/ValidParentheses.java
package com.kmdgnyu;

import java.util.Arrays;
import java.util.List;
import java.util.Stack;

public class ValidParentheses {
    public static boolean isValid(String s) {
        return assemble(s);
    }

    private static boolean assemble(String s) {
        List<Character> open = Arrays.asList('(', '{', '[');
        List<Character> close = Arrays.asList(')', '}', ']');
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (open.contains(s.charAt(i))) {
                stack.push(s.charAt(i));
            } else if (close.contains(s.charAt(i))) {
                char pair = open.get(close.indexOf(s.charAt(i)));
                if (stack.empty()) {
                    return false;
                }
                char lastPushed = stack.pop();
                if (lastPushed != pair) {
                    return false;
                }

            }
        }

        return stack.empty();
    }
}
